radareorg/cutter
# Breakpoint editor dialog

**Detailed description**

* BreakpointDialog allows configuring
  * breakpoint position (address)
  * type HW or SW
  * HW perrmissions and size
  * condition
  * command
  * trace mode (as currently implemented by r2)
* Breakpoint dialog gets used when
  * Creating new breakpoint from BreakpointWidget
  * Edit breakpoint in breakpoint widget using context menu
  * Edit breakpoint in disassembly using context menu
  * Create advanced breakpoint from dissassembly context menu (basic add/remove swbp without popups is still there) - I am open to suggestion for how to call it
* Use more of R2 C api for managing breakpoints
  * fixes index handling bug in breakpointWidget that would happen if you erased first breakpoint and tried to toggle active state for one of the remaining ones
  * allows handling breakpoints that are not defined by address or other special breakpoints



**Test plan (required)**



* Add breakpoint using dissasembly context menu, make sure initial address is correct
* Edit existing breakpoint using disasembly context menu, change a property, verify it was saved using  console command
* Add breakpoint using Ctrl+F2 shortcut in disassembly
* Add breakpoint using breakpoint widget
* Edit breakopint using breakpointWidget context menu, change a property and verify using console
* Change all the properties in BreakpointDialog make sure changes worked using console
* Edit the breakpoint from previous step, make sure all properties are loaded
* Create module breakpoint see limitations bellow
* Create named breakpoint see limitations bellow

![edit_dialog](https://user-images.githubusercontent.com/7101031/71633815-6054cd00-2c1f-11ea-9b62-a131919ac970.png)
![breakpointWidget](https://user-images.githubusercontent.com/7101031/71633816-6054cd00-2c1f-11ea-9651-1b0589b7687c.png)
![context1](https://user-images.githubusercontent.com/7101031/71633818-60ed6380-2c1f-11ea-82be-76b65d1c292d.png)
![context2](https://user-images.githubusercontent.com/7101031/71633820-60ed6380-2c1f-11ea-949f-e09972b713f5.png)




**Known limitations**
* Hardware breakpoints work with gdb server, not with linux-native backend. Don't know about other debugger backends.
* Command gets executed regardless of condition state. Should be dealt with on r2 side.
* Some of the functionality might not work in ESIL mode
* Named (expression) breakpoint isn't recalculated when restarting executable, it is rebased relative to main module instead
* Module breakpoints aren't recalculated when restarting executable
* Module breakpoints can't be distinguished from regular ones in Cutter, this can be reconsidered after they are fully working on r2 side
* Inconsistent base for module breakpoints
* `dbi-*` and `r_bp_del_index` are broken

**Closing issues**
